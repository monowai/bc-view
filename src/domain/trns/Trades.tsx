import React from "react";
import "../../core/css/styles.sass";
import { Link } from "react-router-dom";
import { useAssetTransactions } from "./hooks";
import { ErrorPage } from "../../core/errors/ErrorPage";
import { useAsset } from "../assets/hooks";
import NumberFormat from "react-number-format";
import { isDone } from "../../core/types/typeUtils";
import { ShowError } from "../../core/errors/ShowError";

export function Trades(portfolioId: string, assetId: string): React.ReactElement {
  const trnsResult = useAssetTransactions(portfolioId, assetId, "trades");
  const assetResult = useAsset(assetId);

  if (isDone(trnsResult) && isDone(assetResult)) {
    if (assetResult.error) {
      return ErrorPage(assetResult.error.stack, assetResult.error.message);
    }
    if (trnsResult.error) {
      return <ShowError {...trnsResult.error} />;
    }

    if (trnsResult.data.length > 0) {
      return (
        <div>
          <nav className="container">
            <div className={"page-title"}>
              <div className={"column page-title subtitle is-6"}>
                {assetResult.data.name}:{assetResult.data.market.code}
              </div>
            </div>
          </nav>
          <div className="page-box is-primary has-background-light">
            <div className="container">
              <table className={"table is-striped is-hoverable"}>
                <thead>
                  <tr>
                    <th>Type</th>
                    <th>Currency</th>
                    <th>Trade Date</th>
                    <th align={"right"}>Quantity</th>
                    <th align={"right"}>Price</th>
                    <th align={"right"}>Amount</th>
                    <th align={"right"}>T/B Rate</th>
                    <th align={"right"}>Cash</th>
                    <th align={"right"}>Tax</th>
                    <th align={"right"}>Charges</th>
                    <th>Action</th>
                  </tr>
                </thead>
                <tbody>
                  {trnsResult.data.map((t) => (
                    <tr key={t.id}>
                      <td>{t.trnType}</td>
                      <td>{t.tradeCurrency.code}</td>
                      <td>{t.tradeDate}</td>
                      <td align={"right"}>
                        <NumberFormat
                          value={t.quantity}
                          displayType={"text"}
                          decimalScale={0}
                          fixedDecimalScale={true}
                          thousandSeparator={true}
                        />
                      </td>
                      <td align={"right"}>
                        <NumberFormat
                          value={t.price}
                          displayType={"text"}
                          decimalScale={2}
                          fixedDecimalScale={true}
                          thousandSeparator={true}
                        />
                      </td>
                      <td align={"right"}>
                        <NumberFormat
                          value={t.tradeAmount}
                          displayType={"text"}
                          decimalScale={2}
                          fixedDecimalScale={true}
                          thousandSeparator={true}
                        />
                      </td>

                      <td align={"right"}>
                        <NumberFormat
                          value={t.tradeBaseRate}
                          displayType={"text"}
                          decimalScale={2}
                          fixedDecimalScale={true}
                          thousandSeparator={true}
                        />
                      </td>
                      <td align={"right"}>
                        <NumberFormat
                          value={t.cashAmount}
                          displayType={"text"}
                          decimalScale={2}
                          fixedDecimalScale={true}
                          thousandSeparator={true}
                        />
                      </td>
                      <td align={"right"}>
                        <NumberFormat
                          value={t.tax}
                          displayType={"text"}
                          decimalScale={2}
                          fixedDecimalScale={true}
                          thousandSeparator={true}
                        />
                      </td>
                      <td align={"right"}>
                        <NumberFormat
                          value={t.fees}
                          displayType={"text"}
                          decimalScale={2}
                          fixedDecimalScale={true}
                          thousandSeparator={true}
                        />
                      </td>
                      <td>
                        <Link className="fa fa-edit" to={`/trns/${t.portfolio.id}/${t.id}`} />
                      </td>
                    </tr>
                  ))}
                </tbody>
              </table>
            </div>
          </div>
        </div>
      );
    }
  }
  return (
    <div id="root" data-testid="loading">
      Loading...
    </div>
  );
}

export default Trades;
