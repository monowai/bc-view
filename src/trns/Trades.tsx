import React from "react";
import "../css/styles.sass";
import { Link } from "react-router-dom";
import { useAssetTransactions } from "./hooks";
import { ErrorPage } from "../errors/ErrorPage";
import { useAsset } from "../assets/hooks";
import NumberFormat from "react-number-format";
import { isDone } from "../types/typeUtils";
import { ShowError } from "../errors/ShowError";

export function Trades(portfolioId: string, assetId: string): React.ReactElement {
  const trnsResult = useAssetTransactions(portfolioId, assetId, "trades");
  const assetResult = useAsset(assetId);

  if (isDone(trnsResult) && isDone(assetResult)) {
    if (assetResult.error) {
      return ErrorPage(assetResult.error.stack, assetResult.error.message);
    }
    if (trnsResult.error) {
      return <ShowError error={trnsResult.error} />;
    }

    if (trnsResult.data.length > 0) {
      return (
        <div>
          <nav className="container">
            <div className={"page-title"}>
              <div className={"column page-title subtitle is-6"}>
                {assetResult.data.name}:{assetResult.data.market.code}
              </div>
            </div>
          </nav>
          <div className="page-box is-primary has-background-light">
            <div className="container">
              <table className={"table is-striped is-hoverable"}>
                <thead>
                  <tr>
                    <th>Type</th>
                    <th>Currency</th>
                    <th>Trade Date</th>
                    <th align={"right"}>Amount</th>
                    <th align={"right"}>Tax</th>
                    <th align={"right"}>Charges</th>
                    <th align={"right"}>Quantity</th>
                    <th align={"right"}>Price</th>
                    <th>Action</th>
                  </tr>
                </thead>
                <tbody>
                  {trnsResult.data.map((t) => (
                    <tr key={t.id}>
                      <td>{t.trnType}</td>
                      <td>{t.tradeCurrency.code}</td>
                      <td>{t.tradeDate}</td>
                      <td align={"right"}>
                        <NumberFormat
                          value={t.tradeAmount}
                          displayType={"text"}
                          decimalScale={2}
                          fixedDecimalScale={true}
                          thousandSeparator={true}
                        />
                      </td>
                      <td align={"right"}>
                        <NumberFormat
                          value={t.tax}
                          displayType={"text"}
                          decimalScale={2}
                          fixedDecimalScale={true}
                          thousandSeparator={true}
                        />
                      </td>
                      <td align={"right"}>
                        <NumberFormat
                          value={t.fees}
                          displayType={"text"}
                          decimalScale={2}
                          fixedDecimalScale={true}
                          thousandSeparator={true}
                        />
                      </td>
                      <td align={"right"}>
                        <NumberFormat
                          value={t.quantity}
                          displayType={"text"}
                          decimalScale={0}
                          fixedDecimalScale={true}
                          thousandSeparator={true}
                        />
                      </td>
                      <td align={"right"}>
                        <NumberFormat
                          value={t.price}
                          displayType={"text"}
                          decimalScale={2}
                          fixedDecimalScale={true}
                          thousandSeparator={true}
                        />
                      </td>
                      <td>
                        <Link className="fa fa-edit" to={`/trns/${t.portfolio.id}/${t.id}`} />
                      </td>
                    </tr>
                  ))}
                </tbody>
              </table>
            </div>
          </div>
        </div>
      );
    }
  }
  return (
    <div id="root" data-testid="loading">
      Loading...
    </div>
  );
}

export default Trades;
