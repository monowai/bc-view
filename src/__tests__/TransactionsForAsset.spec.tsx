import React from "react";
import { cleanup, render, screen, waitForElementToBeRemoved } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import nock from "nock";
import { MemoryRouter } from "react-router";
import Trades from "../trns/Trades";

afterEach(cleanup);
afterAll(nock.cleanAll);

jest.mock("react-i18next", () => ({
  useTranslation: () => ({ t: (key) => key }),
}));

jest.mock("@react-keycloak/ssr", () => ({
  useKeycloak: () => ({
    initialized: true,
    keycloak: { token: "123" },
  }),
}));

nock("http://localhost", {
  reqheaders: {
    authorization: "Bearer 123",
  },
})
  .persist(true)
  .get("/bff/trns/test/asset/alphabet/trades")
  .replyWithFile(200, __dirname + "/__contracts__/trans-for-asset.json", {
    "Access-Control-Allow-Origin": "*",
    "Content-type": "application/json",
  })
  .get("/bff/assets/alphabet")
  .replyWithFile(200, __dirname + "/__contracts__/alphabet.json", {
    "Access-Control-Allow-Origin": "*",
    "Content-type": "application/json",
  });

describe("<Trades />", () => {
  it("trades for asset should match snapshot", async () => {
    const TestTrnForAsset = (): JSX.Element => {
      return Trades("test", "alphabet");
    };

    const { container } = render(
      <MemoryRouter initialEntries={["/"]} keyLength={0}>
        <TestTrnForAsset />
      </MemoryRouter>
    );
    await screen.findByTestId("loading");
    await waitForElementToBeRemoved(() => screen.getByTestId("loading"));
    expect(nock.isDone());
    await screen.findByText("BUY");
    expect(container).toMatchSnapshot();
  });
});
