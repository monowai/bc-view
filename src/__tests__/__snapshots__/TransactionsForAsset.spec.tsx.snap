// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Trades /> trades for asset should match snapshot 1`] = `
<div>
  <div>
    <nav
      class="container"
    >
      <div
        class="page-title"
      >
        <div
          class="column page-title subtitle is-6"
        >
          Alphabet Inc - Ordinary Shares - Class C
          :
          NASDAQ
        </div>
      </div>
    </nav>
    <div
      class="page-box is-primary has-background-light"
    >
      <div
        class="container"
      >
        <table
          class="table is-striped is-hoverable"
        >
          <thead>
            <tr>
              <th>
                Type
              </th>
              <th>
                Currency
              </th>
              <th>
                Trade Date
              </th>
              <th
                align="right"
              >
                Amount
              </th>
              <th
                align="right"
              >
                Tax
              </th>
              <th
                align="right"
              >
                Charges
              </th>
              <th
                align="right"
              >
                Quantity
              </th>
              <th
                align="right"
              >
                Price
              </th>
              <th>
                Action
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                BUY
              </td>
              <td>
                USD
              </td>
              <td>
                2019-01-18
              </td>
              <td
                align="right"
              >
                <span>
                  2,160.01
                </span>
              </td>
              <td
                align="right"
              >
                <span>
                  0.00
                </span>
              </td>
              <td
                align="right"
              >
                <span>
                  0.00
                </span>
              </td>
              <td
                align="right"
              >
                <span>
                  2
                </span>
              </td>
              <td
                align="right"
              >
                <span>
                  1,080.00
                </span>
              </td>
              <td>
                <a
                  class="fa fa-edit"
                  href="/trns/test/68"
                />
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
